src/NuGet.Core/NuGet.Build.Tasks/RestoreTask.cs

```diff
@@ -242,5 +314,243 @@ public void Dispose()
+            var nuGetPackageManager = new NuGetPackageManager(sourceRepositoryProvider, settings, repositoryPath);
+
+            // TODO: different default?  Allow user to specify?
+            var packageSaveMode = Packaging.PackageSaveMode.Defaultv2;
+
+            var missingPackageReferences = installedPackageReferences.Where(reference =>
+                !nuGetPackageManager.PackageExistsInPackagesFolder(reference.PackageIdentity, packageSaveMode)).ToArray();
+
+            if (missingPackageReferences.Length == 0)
+            {
+                return new RestoreSummary(true);
+            }
+            var packageRestoreData = missingPackageReferences.Select(reference =>
+                new PackageRestoreData(
+                    reference,
+                    new[] { firstPackagesConfigPath },
+                    isMissing: true));
+
+            var repositories = packageSources
+                .Select(sourceRepositoryProvider.CreateRepository)
+                .ToArray();
+
+            var installCount = 0;
+            var failedEvents = new ConcurrentQueue<PackageRestoreFailedEventArgs>();
+            var collectorLogger = new RestoreCollectorLogger(new MSBuildLogger(Log));
+
+            var packageRestoreContext = new PackageRestoreContext(
+                nuGetPackageManager,
+                packageRestoreData,
+                CancellationToken.None,
+                packageRestoredEvent: (sender, args) => { Interlocked.Add(ref installCount, args.Restored ? 1 : 0); },
+                packageRestoreFailedEvent: (sender, args) => { failedEvents.Enqueue(args); },
+                sourceRepositories: repositories,
+                maxNumberOfParallelTasks: RestoreDisableParallel
+                    ? 1
+                    : PackageManagementConstants.DefaultMaxDegreeOfParallelism,
+                logger: collectorLogger);
+
+            // TODO: Check require consent?
+            // CheckRequireConsent();
+
+            var clientPolicyContext = ClientPolicyContext.GetClientPolicy(settings, collectorLogger);
+            var projectContext = new ConsoleProjectContext(collectorLogger)
+            {
+                PackageExtractionContext = new PackageExtractionContext(
+                    Packaging.PackageSaveMode.Defaultv2,
+                    PackageExtractionBehavior.XmlDocFileSaveMode,
+                    clientPolicyContext,
+                    collectorLogger)
+            };
+
+            // if (EffectivePackageSaveMode != Packaging.PackageSaveMode.None)
+            {
+                projectContext.PackageExtractionContext.PackageSaveMode = packageSaveMode;
+            }
+
+            using (var cacheContext = new SourceCacheContext())
+            {
+                cacheContext.NoCache = RestoreNoCache;
+
+                // TODO: Direct download?
+                // //cacheContext.DirectDownload = DirectDownload;
```

Comment:
There's no DirectDownload parameter to `RestoreTask` so I figured I'd not allow it for this scenario.  Just making sure you agree.